chainerrl/experiments/train_agent_batch.py

```diff
@@ -0,0 +1,222 @@
+
+
+def _get_mask(num_processes, info, done, episode_len, max_episode_len):
+    if max_episode_len is None:
+        reset_mask = np.zeros(num_processes, dtype=bool)
+    else:
+        reset_mask = episode_len == max_episode_len
+
+    for i, reset in zip(info, reset_mask):
+        i['reset'] = reset
+
+    mask = np.logical_not(np.logical_or(reset_mask, done))
+    return info, mask
+
+
+def _get_obs(env, obs, mask, num_processes):
+    return np.vstack([env.reset()[i] if not mask[i] else obs[i] for i in range(num_processes)])  # NOQA
+
+
+def _write_to_file(outdir, step, avg_r):
+    with open(os.path.join(outdir, 'training_r.txt'), 'a+') as f:
+        print('\t'.join(str(x) for x in [step, avg_r]), file=f)
+
+
+def train_agent_batch(agent, env, steps, outdir, log_interval=None,
+                      max_episode_len=None, eval_interval=None,
+                      step_offset=0, evaluator=None, successful_score=None,
+                      step_hooks=[], deque_maxlen=100, logger=None,
+                      save_training_r=False):
+
+    logger = logger or logging.getLogger(__name__)
+    d = deque(maxlen=deque_maxlen)
+
+    try:
+        num_processes = env.num_envs
+    except AttributeError:
+        logger.error('Please pass a VectorEnv instance. \
+You passed: {}'.format(type(env)))
+        raise
+
+    episode_r = np.zeros(num_processes, dtype='f')
+    episode_idx = np.zeros(num_processes, dtype='f')
+    episode_len = np.zeros(num_processes, dtype='f')
+
+    # o_0, r_0
+    obs = env.reset()
+    r = np.zeros(num_processes, dtype='f')
+
+    t = step_offset
+    if hasattr(agent, 't'):
+        agent.t = step_offset
+
+    try:
+        while t < steps:
+            # a_t
+            action = agent.batch_act_and_train(obs)
+            # o_{t+1}, r_{t+1}
+            obs, r, done, info = env.step(action)
+            # Compute mask for done and reset
+            if max_episode_len is None:
+                reset_mask = np.zeros(num_processes, dtype=bool)
+            else:
+                reset_mask = episode_len == max_episode_len
+            # Package mask inside info dict
+            for i, reset in zip(info, reset_mask):
+                i['reset'] = reset
+            # Make mask. 0 if done/reset, 1 if pass
+            masks = np.logical_not(np.logical_or(reset_mask, done))
+            # Reset environment whenever done/reset
+            obs_ = _get_obs(env, obs, masks, num_processes)
+            # Train agent
+            agent.batch_observe_and_train(obs_, r, done, info)
+            # Update reward for current episode
+            episode_r += r
+            episode_len += 1
+            episode_idx += 1 - masks
+            # Add to deque whenever done/reset
+            episode_r_ = np.ma.masked_array(episode_r, masks)
+            d.extend(episode_r_.compressed())
+            # Start new episode for those with mask
+            episode_r *= masks
+            episode_len *= masks
+            t += 1
```

Comment:
`t` should be the total number of transitions experienced so far. So, it should increase by `num_envs`, not by 1. By doing so, we can keep the other hyperparameters and only change `num_envs` to trade cpus with computation time.